lib/gc/ExecutionEngine/CPURuntime/Microkernel/BrgemmOnednn.cpp

```diff
@@ -0,0 +1,159 @@
+static thread_local char scratch[SCRATCH_SIZE] = {0};
+
+extern "C" {
+
+int64_t dnnl_brgemm_dispatch(int64_t M, int64_t N, int64_t K, int64_t LDA,
+                             int64_t LDB, int64_t LDC, int64_t stride_a,
+                             int64_t stride_b, float beta, int64_t dtypeA,
+                             int64_t dtypeB) {
+  brgemm_desc_t desc;
+  brgemm_kernel_t *kernel;
+
+  auto dnnl_dtypeA = static_cast<dnnl_data_type_t>(dtypeA);
+  auto dnnl_dtypeB = static_cast<dnnl_data_type_t>(dtypeB);
+  int64_t dtypeA_size = dnnl::impl::types::data_type_size(dnnl_dtypeA);
+  int64_t dtypeB_size = dnnl::impl::types::data_type_size(dnnl_dtypeB);
+  brgemm_strides_t stride_info{stride_a * dtypeA_size, stride_b * dtypeB_size};
+
+  dnnl::impl::status_t status = brgemm_desc_init(
+      &desc, cpu_isa_t::isa_undef, brgemm_batch_kind_t::brgemm_strd,
+      dnnl_dtypeA, dnnl_dtypeB, /*transA=*/false, /*transB=*/false,
+      brgemm_layout_t::brgemm_row_major, 1.0f, beta, LDA, LDB, LDC, M, N, K,
+      &stride_info);
+  assert(status == dnnl::impl::status::success &&
+         "Failed to initialize BRGEMM descriptor");
+
+  status = brgemm_kernel_create(&kernel, desc);
+  assert(status == dnnl::impl::status::success &&
+         "Failed to JIT BRGEMM kernel");
+
+  brgemm_attr_t dnnl_attrs;
+  brgemm_desc_set_attr(&desc, dnnl_attrs);
+
+  // TODO(haixin): Reuse identical palettes across kernels
+  char *palette_buffer = nullptr;
+  if (desc.is_tmm) {
+    palette_buffer = new char[PALETTE_SIZE];
+    dnnl::impl::status_t status = brgemm_init_tiles(desc, palette_buffer);
+    assert(status == dnnl::impl::status::success &&
+           "Failed to initialize palette for BRGEMM");
+  }
+
+  std::lock_guard<std::mutex> g(g_brgemm_mutex);
+  g_brgemm_desc_list.push_back(desc);
+  g_brgemm_kernel_list.push_back(kernel);
+  g_brgemm_palette.emplace_back(palette_buffer);
+
+  return g_brgemm_desc_list.size() - 1;
+}
+
+void dnnl_brgemm_tileconfig(int64_t kernel_idx) {
+  char *palette_buffer = nullptr;
+  {
+    std::lock_guard<std::mutex> g(g_brgemm_mutex);
+    assert(kernel_idx >= 0 && kernel_idx < (int64_t)g_brgemm_desc_list.size() &&
+           "Invalid kernel handler");
+    brgemm_desc_t &desc = g_brgemm_desc_list[kernel_idx];
+    if (!desc.is_tmm) {
+      return;
+    }
+    palette_buffer = g_brgemm_palette[kernel_idx].get();
+  }
+
+  assert(palette_buffer != nullptr && "Invalid palette for BRGEMM kernel");
+  amx_tile_configure(palette_buffer);
+}
+
+void dnnl_brgemm_tilerelease() {
+  if (!mayiuse(avx512_core_amx)) {
+    return;
+  }
+
+  amx_tile_release();
+}
+
+void dnnl_brgemm_execute(int64_t kernel_idx, void *A, uint64_t A_offset,
+                         void *B, uint64_t B_offset, void *C, uint64_t C_offset,
+                         int num) {
+  brgemm_kernel_t *kernel = nullptr;
+  size_t A_offset_in_bytes;
+  size_t B_offset_in_bytes;
+  size_t C_offset_in_bytes;
+  {
+    std::lock_guard<std::mutex> g(g_brgemm_mutex);
```

Comment:
It would be better to use a [RWLock](https://github.com/llvm/llvm-project/blob/67cb04035fe831a3b5df98dabc44b53ba5ff7126/llvm/include/llvm/Support/RWMutex.h#L99)/[shared_lock](https://en.cppreference.com/w/cpp/thread/shared_lock) here so that other readers could access these data at the same time

Reply:
Good suggestion. Changed to STL based RW lock implementation.